C++ API parity: at::Tensor::data

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26150 C++ API parity: at::Tensor::grad
* **#26008 C++ API parity: at::Tensor::data**

Differential Revision: [D17343488](https://our.internmc.facebook.com/intern/diff/D17343488)

pbelevich committed:
C++ API parity: at::Tensor::data

yf225 commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -295,12 +295,6 @@ class CAFFE2_API Tensor {
-  T * data() const {
-    return data_ptr<T>();
-  }
-
```
ditto here

yf225 commented on aten/src/ATen/test/quantized_test.cpp:
```diff
@@ -131,14 +131,14 @@ TEST(TestQTensor, EmptyPerchannelQuantized) {
       at::device(at::kCPU).dtype(kQUInt8));
   // Assigning to QTensor
-  auto* q_data = q.data<quint8>();
+  auto* q_data = static_cast<quint8*>(q.data_ptr());
```
I think we can use `data_ptr<T>` here:
```suggestion
  auto* q_data = q.data_ptr<quint8>();
```

yf225 commented on aten/src/ATen/test/quantized_test.cpp:
```diff
@@ -131,14 +131,14 @@ TEST(TestQTensor, EmptyPerchannelQuantized) {
   // dequantize
   auto r = q.dequantize();
-  auto* r_data = r.data<float>();
+  auto* r_data = static_cast<float*>(r.data_ptr());
```
ditto - I think we can use `data_ptr<T>` here:
```suggestion
  auto* r_data = r.data_ptr<float>();
```

yf225 commented on test/cpp/api/tensor.cpp:
```diff
@@ -311,3 +311,8 @@ TEST(TensorTest, DataPtr) {
+
+TEST(TensorTest, Data) {
+  auto tensor = torch::ones({3, 3}, torch::requires_grad());
+  auto data = tensor.data();
```
It would be awesome to add `ASSERT_TRUE(torch::equal(tensor, data))` here :3

yf225 reviewed (commented):
Looks awesome! Thanks a lot @pbelevich :D

pbelevich committed:
Update on "C++ API parity: at::Tensor::data"

yf225 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -35,6 +35,9 @@
 
+- func: data(Tensor self) -> Tensor
+  variants: method
+
```
To fix the CI errors, we likely need to add `data` to `SKIP_PYTHON_BINDINGS` in `tools/autograd/gen_python_functions.py`.

yf225 reviewed (approved):
Looks great, thanks @pbelevich ! We should be able to merge it after the fix for `.data` Python binding.

pbelevich committed:
Update on "C++ API parity: at::Tensor::data"

pbelevich committed:
Update on "C++ API parity: at::Tensor::data"


Differential Revision: [D17343488](https://our.internmc.facebook.com/intern/diff/D17343488)

pbelevich committed:
Update on "C++ API parity: at::Tensor::data"


Differential Revision: [D17343488](https://our.internmc.facebook.com/intern/diff/D17343488)

pbelevich committed:
Update on "C++ API parity: at::Tensor::data"


Differential Revision: [D17343488](https://our.internmc.facebook.com/intern/diff/D17343488)

pbelevich committed:
Update on "C++ API parity: at::Tensor::data"


Differential Revision: [D17343488](https://our.internmc.facebook.com/intern/diff/D17343488)
